pkg/plugins/resource/resource.go

```diff
@@ -0,0 +1,171 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	clusterapiv1 "open-cluster-management.io/api/cluster/v1"
+	clusterapiv1alpha1 "open-cluster-management.io/api/cluster/v1alpha1"
+	"open-cluster-management.io/placement/pkg/plugins"
+)
+
+const (
+	placementLabel = "cluster.open-cluster-management.io/placement"
+	description    = `
+	ResourceRatio[ResourceType] and ResourceAllocatable[ResourceType] prioritizer makes the scheduling 
+	decisions based on the resource allocatable to capacity ratio or allocatable of managed clusters. 
+	The [ResourceType] could be CPU or Memory.
+	The clusters that has the most allocatable to capacity ratio or allocatable are given the highest score, 
+	while the least is given the lowest score.
+	`
+)
+
+var _ plugins.Prioritizer = &Resource{}
+
+type Resource struct {
+	handle   plugins.Handle
+	resource string
+	mode     string
+}
+
+type ResourceBuilder struct {
+	resource *Resource
+}
+
+func NewResourceBuilder(handle plugins.Handle) *ResourceBuilder {
+	return &ResourceBuilder{
+		resource: &Resource{
+			handle: handle,
+		},
+	}
+}
+
+func (r *ResourceBuilder) WithResource(resource string) *ResourceBuilder {
+	r.resource.resource = resource
+	return r
+}
+
+func (r *ResourceBuilder) WithMode(mode string) *ResourceBuilder {
+	r.resource.mode = mode
+	return r
+}
+
+func (r *ResourceBuilder) Build() *Resource {
+	return r.resource
+}
+
+func (r *Resource) Name() string {
+	return "Resource" + r.mode + r.resource
+}
+
+func (r *Resource) Description() string {
+	return description
+}
+
+func (r *Resource) Score(ctx context.Context, placement *clusterapiv1alpha1.Placement, clusters []*clusterapiv1.ManagedCluster) (map[string]int64, error) {
+	f := r.mode + r.resource
+	switch {
+	case f == "RatioMemory":
```

Comment:
with above thing, what you need to do here is

```
switch mode {
case "Ratio":
    return mostResourceRatioScores(c.resource, clusters)
case "Allocatable":
    return mostResourceAllocatableScores(c.resource, clusters)
}
```

Reply:
Updated.